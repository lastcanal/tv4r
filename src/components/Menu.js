import React, { useEffect } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { withStyles } from '@material-ui/core/styles'
import Container from '@material-ui/core/Container'
import ToolBar from '@material-ui/core/ToolBar'
import Paper from '@material-ui/core/Paper'
import Grid from '@material-ui/core/Grid'

import {
  selectSubreddit,
  nextPost,
  previousPost,
} from '../actions'

import Posts from './Posts'
import Picker from './Picker'
import Controls from './Controls'
import Title from './Title'

const styles = ({ palette }) => ({
  root: {
    position: 'fixed',
    left: 0,
    bottom: 0,
    padding: 0,
    width: '100%',
    backgroundColor: palette.background.default,
  },
})

const Menu = ({ classes, dispatch, post, posts, selectedSubreddit, subreddits }) => {

  const changeSubreddit = ({ value }) => {
    dispatch(selectSubreddit(value))
  }

  useEffect(() => {
    const handleKeyDown = (e) => {
      switch (e.key) {
        case 'ArrowRight':
          return dispatch(nextPost(posts))
        case 'ArrowLeft':
          return dispatch(previousPost(posts))
        default:
          return
      }
    }

    document.body.addEventListener('keydown', handleKeyDown)
    return () => {
      document.body.removeEventListener('keydown', handleKeyDown)
    }
  }, [posts])

  return (
    <Paper>
      <Container classes={classes} maxWidth={false}>
        <ToolBar>
          <Grid container spacing={3}>
            <Grid item xs={12}>
              <Picker
                value={selectedSubreddit}
                onChange={changeSubreddit}
                options={subreddits}
              />
            </Grid>
          </Grid>
        </ToolBar>
        <ToolBar>
          <Posts />
        </ToolBar>
        <ToolBar>
          <Controls
            dispatch={dispatch}
            posts={posts}
            selectedSubreddit={selectedSubreddit}
          />
          <Title post={post} />
        </ToolBar>
      </Container>
    </Paper>
  )

}

Menu.propTypes = {
  selectedSubreddit: PropTypes.string,
  posts: PropTypes.array,
  post: PropTypes.object,
  isFetching: PropTypes.bool.isRequired,
  lastUpdated: PropTypes.number,
  subreddits: PropTypes.array,
  dispatch: PropTypes.func.isRequired,
  classes: PropTypes.object.isRequired,
}

const mapStateToProps = state => {
  const { selectedSubreddit, postsBySubreddit, subreddits } = state
  const selectedPost = postsBySubreddit.cursor || {}
  const { isFetching, lastUpdated, items: posts } = postsBySubreddit[
    selectedSubreddit
  ] || {
    isFetching: true,
    items: [],
  }

  return {
    postsBySubreddit,
    selectedSubreddit,
    posts,
    isFetching,
    post: selectedPost.post,
    lastUpdated,
    subreddits,
  }
}

export default connect(mapStateToProps)(withStyles(styles)(Menu))
