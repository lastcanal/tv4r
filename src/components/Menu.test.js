import { Provider } from 'react-redux'

import Menu from './Menu'

describe('menu', () => {
  it('should render empty menu', () => {
    const wrapper = mount(
      <Provider store={makeStore()}>
        <Menu />
      </Provider>
    )

    expect(wrapper).toMatchSnapshot()
  })
})

test('should render new subreddit', () => {
  const store = makeStore({
    postsBySubreddit: {
      foo: {
        scope: 'hot',
        hot: [],
        isFetching: false,
        didInvalidate: false,
      },
      bar: {
        scope: 'hot',
        hot: [],
        isFetching: false,
        didInvalidate: false,
      },
    },
    selectedSubreddit: 'foo',
    selectedPost: { post: {} },
  })

  const wrapper = mount(
    <Provider store={store}>
      <Menu />
    </Provider>
  )

  wrapper.setProps({ selectedSubreddit: 'bar' })

  wrapper.update()

  expect(wrapper).toMatchSnapshot()
})
