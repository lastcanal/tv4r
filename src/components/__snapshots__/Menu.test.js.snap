// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should render empty menu 1`] = `ReactWrapper {}`;

exports[`should render new subreddit 1`] = `ReactWrapper {}`;
